From 50b4d1c13ae64a8459e23e81ff2c0fbe26ff3276 Mon Sep 17 00:00:00 2001
From: Ramax Lo <ramax.lo@mediatek.com>
Date: Thu, 9 Mar 2023 14:48:14 +0800
Subject: [PATCH] cmd: Add new command to source embedded script

Add new command 'bootembedded' to run the script appended to the end of
u-boot binary.

Also add new env 'bootcmd_embedded' for calling the command.

Signed-off-by: Ramax Lo <ramax.lo@mediatek.com>
---
 cmd/Makefile                    |  1 +
 cmd/bootembedded.c              | 38 +++++++++++++++++++++++++++++++++
 include/config_distro_bootcmd.h |  1 +
 3 files changed, 40 insertions(+)
 create mode 100644 cmd/bootembedded.c

diff --git a/cmd/Makefile b/cmd/Makefile
index 6e87522b62..cb5e63e6eb 100644
--- a/cmd/Makefile
+++ b/cmd/Makefile
@@ -10,6 +10,7 @@ obj-$(CONFIG_CMD_BOOTM) += bootm.o
 obj-y += help.o
 obj-y += panic.o
 obj-y += version.o
+obj-y += bootembedded.o
 
 # command
 obj-$(CONFIG_CMD_ACPI) += acpi.o
diff --git a/cmd/bootembedded.c b/cmd/bootembedded.c
new file mode 100644
index 0000000000..c4c191b542
--- /dev/null
+++ b/cmd/bootembedded.c
@@ -0,0 +1,38 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Copyright 2023 MediaTek, Inc.
+ */
+
+#include <common.h>
+#include <command.h>
+#include <linux/libfdt.h>
+
+extern uint8_t _start;
+extern uint8_t _end;
+extern uint32_t _TEXT_BASE;
+static int do_bootembedded(struct cmd_tbl *cmdtp, int flag, int argc,
+		      char *const argv[])
+{
+	uint8_t *sp = (uint8_t *)&_start;
+	uint8_t *ep = (uint8_t *)&_end;
+	uint8_t *text_end = (uint8_t *)(_TEXT_BASE + (ep - sp));
+	uint32_t fdtsz = fdt_totalsize(text_end);
+	uint8_t *script_start = text_end + fdtsz;
+	int8_t cmd[64];
+
+	printf("start: 0x%x, end: 0x%x\n", sp, ep);
+	printf("text_start: 0x%x, text_end: 0x%x\n", _TEXT_BASE, text_end);
+	printf("fdtsz: 0x%x (%d)\n", fdtsz, fdtsz);
+	printf("script: 0x%x\n", script_start);
+
+	sprintf(cmd, "source %p", script_start);
+	printf("run '%s'\n", cmd);
+
+	return run_command_list(cmd, -1, 0);
+}
+
+U_BOOT_CMD(
+	bootembedded,	1,		1,	do_bootembedded,
+	"Run embedded script appended to the end of u-boot binary",
+	""
+);
diff --git a/include/config_distro_bootcmd.h b/include/config_distro_bootcmd.h
index 5506f3168f..95d5997fb8 100644
--- a/include/config_distro_bootcmd.h
+++ b/include/config_distro_bootcmd.h
@@ -496,6 +496,7 @@
 		"setenv devplist\0"					  \
 	\
 	BOOT_TARGET_DEVICES(BOOTENV_DEV)                                  \
+	"bootcmd_embedded=bootembedded\0"                                 \
 	\
 	"distro_bootcmd=" BOOTENV_SET_SCSI_NEED_INIT                      \
 		BOOTENV_SET_NVME_NEED_INIT                                \
-- 
2.18.0

